File: javaScript/reverse_shuffle_merge.js
function reverseShuffleMerge(s) {
    // Frequency array for 26 lowercase letters
    const freq = new Array(26).fill(0);
    for (let i = 0; i < s.length; i++) {
        freq[s.charCodeAt(i) - 97]++;
    }
    
    // Need half of each character
    const need = new Array(26);
    for (let i = 0; i < 26; i++) {
        need[i] = freq[i] / 2;
    }
    
    // Track used and remaining
    const used = new Array(26).fill(0);
    const remaining = [...freq];
    
    // Stack for result
    const stack = [];
    
    // Traverse from end
    for (let i = s.length - 1; i >= 0; i--) {
        const ch = s[i];
        const idx = ch.charCodeAt(0) - 97;
        remaining[idx]--;
        
        if (used[idx] >= need[idx]) {
            continue;
        }
        
        while (stack.length > 0) {
            const top = stack[stack.length - 1];
            const topIdx = top.charCodeAt(0) - 97;
            if (top > ch && remaining[topIdx] > need[topIdx] - used[topIdx]) {
                stack.pop();
                used[topIdx]--;
            } else {
                break;
            }
        }
        
        stack.push(ch);
        used[idx]++;
    }
    
    return stack.join('');
}

File: cpp/reverse_shuffle_merge.cpp
string reverseShuffleMerge(string s) {
    int n = s.length();
    
    // Count frequency of each character in s
    vector<int> freq(26, 0);
    for (char c : s) {
        freq[c - 'a']++;
    }
    
    // Each character in answer appears half the times
    vector<int> need(26, 0);
    for (int i = 0; i < 26; i++) {
        need[i] = freq[i] / 2;
    }
    
    // Stack to build the answer
    vector<char> result;
    
    // Track remaining characters and used characters
    vector<int> remaining(26, 0);
    remaining = freq;
    
    vector<int> used(26, 0);
    
    // Traverse from the end (reverse of s)
    for (int i = n - 1; i >= 0; i--) {
        char ch = s[i];
        int idx = ch - 'a';
        
        // Decrement remaining count
        remaining[idx]--;
        
        // If we already have enough of this character in answer, skip
        if (used[idx] >= need[idx]) {
            continue;
        }
        
        // Greedy: pop from stack if we can get a smaller character
        while (!result.empty() && 
               result.back() > ch && 
               remaining[result.back() - 'a'] > need[result.back() - 'a'] - used[result.back() - 'a']) {
            used[result.back() - 'a']--;
            result.pop_back();
        }
        
        // Add current character to result
        result.push_back(ch);
        used[idx]++;
    }
    
    // Convert result to string
    string answer(result.begin(), result.end());
    return answer;
}